percona/controllers/pgc/secrets.go

```diff
@@ -0,0 +1,297 @@
+		if err != nil {
+			return errors.Wrap(err, "create users secret")
+		}
+		secretsData = generatedSecretsData
+	} else {
+		generatedSecretsData, err := c.generateUsersInternalSecretsDataFromSecret(usersSecret, clusterName)
+		if err != nil {
+			return errors.Wrap(err, "generate users internal data from secret")
+		}
+		secretsData = generatedSecretsData
+	}
+
+	err = c.createUsersInternalSecrets(secretsData, namespace, clusterName)
+	if err != nil {
+		return errors.Wrap(err, "create users internal secrets")
+	}
+
+	return nil
+}
+
+func (c *Controller) createUsersSecret(secretsData []SecretData, secretName, namespace string) error {
+	ctx := context.TODO()
+	data := make(map[string][]byte)
+	for _, secret := range secretsData {
+		data[secret.Data.Name] = []byte(secret.Data.Password)
+	}
+	secretData, err := json.Marshal(data)
+	if err != nil {
+		return errors.Wrap(err, "marshal users secret data")
+	}
+	secretDataHash := sha256Hash(secretData)
+	s := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: namespace,
+			Annotations: map[string]string{
+				"last-applied-secret": secretDataHash,
+			},
+		},
+		Data: data,
+	}
+	_, err = c.Client.CoreV1().Secrets(namespace).Create(ctx, s, metav1.CreateOptions{})
+	if err != nil && !kerrors.IsAlreadyExists(err) {
+		return errors.Wrapf(err, "create secret %s", secretName)
+	}
+
+	return nil
+}
+
+func (c *Controller) generateUsersInternalSecretsDataFromSecret(usersSecret *v1.Secret, clusterName string) ([]SecretData, error) {
+	secrets := []SecretData{}
+	for userName, password := range usersSecret.Data {
+		secretData, err := generateUserSecretData(userName, clusterName, string(password))
+		if err != nil {
+			return nil, errors.Wrapf(err, "get secret data for %s", userName)
+		}
+		secrets = append(secrets, secretData)
+	}
+
+	return secrets, nil
+}
+
+func (c *Controller) GenerateUsersInternalSecretsData(clusterName string) ([]SecretData, error) {
+	bouncerUserData, err := generateUserSecretData("pgbouncer", clusterName, "")
+	if err != nil {
+		return nil, errors.Wrap(err, "generate pgbouncer")
+	}
+	postgresUserData, err := generateUserSecretData("postgres", clusterName, "")
+	if err != nil {
+		return nil, errors.Wrap(err, "generate postgres")
+	}
+	primaryUserData, err := generateUserSecretData("primaryuser", clusterName, "")
+	if err != nil {
+		return nil, errors.Wrap(err, "generate primaryuser")
+	}
+
+	secrets := []SecretData{
+		bouncerUserData,
+		postgresUserData,
+		primaryUserData,
+	}
+
+	return secrets, nil
+}
+
+func generateUserSecretData(userName, clusterName, password string) (SecretData, error) {
+	var secretName, roles, usersTXT string
+	if len(password) == 0 {
+		generatedPassword, err := util.GeneratePassword(util.DefaultGeneratedPasswordLength)
+		if err != nil {
+			return SecretData{}, errors.Wrap(err, "generate password")
+		}
+		password = generatedPassword
+	}
+
+	switch userName {
+	case "postgres":
```

Comment:
also we need to have possibility of changing password for  user which was created from CR https://github.com/percona/percona-postgresql-operator/blob/main/deploy/cr.yaml#L17